src/typechecker.erl

```diff
@@ -942,55 +942,58 @@ type_check_lc(Env, Expr, [{generate, _, Pat, Gen} | Quals]) ->
 
 
 
+type_check_expr_in(Env, ResTy, Expr) ->
+    NormResTy = normalize(ResTy, Env#env.tenv),
+    do_type_check_expr_in(Env, NormResTy, Expr).
```

Comment:
This is good, except the type_error will contain the normalized type. It is better to do it in one place than in all the `*_in` functions. The original type can easily be restored with using the try-catch and re-throw method I sketched earlier (although using get_stacktrace/0 for pre-OTP-21), at least temporarily until we find a better solution. As a rule for this, let's say that type to replace is always the last element in the tuple if the first element is `type_error`.

Reply:
ok, I will add another commit with re-throw and assuming last type_error element is the one to replace

Reply:
Just noting: keeping the original type has the funny side-effect of error messages like this:
```
The operator 'andalso' on line 7 is given a non-boolean argument of type bool()
```
whereas showing the normalized type better reveals the problem (`type_check_logic_op_in` does not handle union as boolean())
```
The operator 'andalso' on line 7 is given a non-boolean argument of type true | false
```

This can be remedied by printing the whole type stack as you proposed.